src/Framework/Framework/Compilation/ViewCompiler/DefaultViewCompilerCodeEmitter.cs

```diff
@@ -0,0 +1,360 @@
+
+            blockStack.Peek().Expressions.Add(assigment);
+        }
+
+        /// <summary>
+        /// Emits the add directive.
+        /// </summary>
+        public void EmitAddDirective(string controlName, string name, string value)
+        {
+            EmitAddToDictionary(controlName, "Directives", name, EmitValue(value));
+        }
+
+        public ParameterExpression EmitEnsureCollectionInitialized(string parentName, DotvvmProperty property)
+        {
+            //if([parentName].GetValue(property) == null)
+            //{
+            //  [parentName].SetValue(property, new [property.PropertyType]());
+            //}
+
+            var parentParameter = GetParameterOrVariable(parentName);
+
+            var getPropertyValue = Expression.Call(
+                parentParameter,
+                "GetValue",
+                emptyTypeArguments,
+                /*property:*/ CreateDotvvmPropertyIdentifier(property),
+                /*inherit:*/ EmitValue(true /*default*/ ));
+
+            var ifCondition = Expression.Equal(getPropertyValue, Expression.Constant(null));
+            var statement = Expression.Call(parentParameter, "SetValue", emptyTypeArguments, CreateDotvvmPropertyIdentifier(property), EmitCreateObjectExpression(property.PropertyType, new Expression[] { }));
+            var ifStatement = Expression.IfThen(ifCondition, statement);
+
+            blockStack.Peek().Expressions.Add(ifStatement);
+
+            //var c = ([property.PropertyType])[parentName].GetValue(property);
+
+            return EmitCreateVariable(Expression.Convert(getPropertyValue, property.PropertyType));
+        }
+
+        /// <summary>
+        /// Emits the return clause.
+        /// </summary>
+        public void EmitReturnClause(string variableName)
+        {
+            var parameter = GetParameterOrVariable(variableName);
+            blockStack.Peek().Expressions.Add(parameter);
+        }
+
+        public ParameterExpression GetParameterOrVariable(string identifierName)
+            => GetCurrentBlock().GetParameterOrVariable(identifierName);
+
+        private BlockInfo GetCurrentBlock() => blockStack.Peek();
+        public ParameterExpression EmitParameter(string name, Type type) => Expression.Parameter(type, name);
+
+        public ParameterExpression[] EmitControlBuilderParameters()
+            => new[]
+            {
+                EmitParameter(ControlBuilderFactoryParameterName, typeof(IControlBuilderFactory)),
+                EmitParameter(ServiceProviderParameterName, typeof(IServiceProvider))
+            };
+
+        /// <summary>
+        /// Pushes the new method.
+        /// </summary>
+        public void PushNewMethod(string name, Type returnType, params ParameterExpression[] parameters)
+        {
+            blockStack.Push(new BlockInfo(name, returnType, parameters));
+        }
+
+        /// <summary>
+        /// Pops the method.
+        /// </summary>
+        public Delegate PopMethod()
```

Comment:
```suggestion
        public T PopMethod<T>() where T: Delegate
```

The type of the lambda that we want is know at both places where this is invoked, so I think we could make it like this

Reply:
Solved